python/pyspark/mllib/linalg/distributed.py

```diff
@@ -0,0 +1,513 @@
+    ...     mat._java_matrix_wrapper._java_model)
+    False
+
+    >>> mat_same = _create_from_java(mat._java_matrix_wrapper._java_model)
+    >>> (mat_same._java_matrix_wrapper._java_model ==
+    ...     mat._java_matrix_wrapper._java_model)
+    True
+    """
+    if isinstance(java_matrix, JavaObject):
+        class_name = java_matrix.getClass().getSimpleName()
+        if class_name == "RowMatrix":
+            rows = callJavaFunc(SparkContext._active_spark_context, getattr(java_matrix, "rows"))
+            return RowMatrix(rows, java_matrix=java_matrix)
+        elif class_name == "IndexedRowMatrix":
+            # We use DataFrames for serialization of IndexedRows from
+            # Java, so we first convert the RDD of rows to a DataFrame
+            # on the Scala/Java side. Then we map each Row in the
+            # DataFrame back to an IndexedRow on this side.
+            rows_df = callMLlibFunc("getIndexedRows", java_matrix)
+            rows = rows_df.map(lambda row: IndexedRow(row[0], row[1]))
+            return IndexedRowMatrix(rows, java_matrix=java_matrix)
+        elif class_name == "CoordinateMatrix":
+            # We use DataFrames for serialization of MatrixEntry entries
+            # from Java, so we first convert the RDD of entries to a
+            # DataFrame on the Scala/Java side. Then we map each Row in
+            # the DataFrame back to a MatrixEntry on this side.
+            entries_df = callMLlibFunc("getMatrixEntries", java_matrix)
+            entries = entries_df.map(lambda row: MatrixEntry(row[0], row[1], row[2]))
+            return CoordinateMatrix(entries, java_matrix=java_matrix)
+        else:
+            raise TypeError("Cannot create distributed matrix from Java %s" % class_name)
+    else:
+        raise TypeError("java_matrix should be JavaObject, got %s" % type(java_matrix))
+
+
+class DistributedMatrix(object):
+    """
+    Represents a distributively stored matrix backed by one or
+    more RDDs.
+
+    """
+    def numRows(self):
+        """Get or compute the number of rows."""
+        raise NotImplementedError
+
+    def numCols(self):
+        """Get or compute the number of cols."""
+        raise NotImplementedError
+
+
+class RowMatrix(DistributedMatrix):
+    """
+    .. note:: Experimental
+
+    Represents a row-oriented distributed Matrix with no meaningful
+    row indices.
+
+    :param rows: An RDD of vectors.
+    :param numRows: Number of rows in the matrix. A non-positive
+                    value means unknown, at which point the number
+                    of rows will be determined by the number of
+                    records in the `rows` RDD.
+    :param numCols: Number of columns in the matrix. A non-positive
+                    value means unknown, at which point the number
+                    of columns will be determined by the size of
+                    the first row.
+    """
+    def __init__(self, rows, numRows=0, numCols=0, java_matrix=None):
```

Comment:
Yeah, I think that would be really clean, so long as it's documented well internally, as @MechCoder mentioned.